modules/calib3d/src/ptsetreg.cpp

```diff
@@ -501,6 +501,242 @@ class Affine3DEstimatorCallback : public PointSetRegistrator::Callback
     }
 };
 
+class Affine2DEstimatorCallback : public PointSetRegistrator::Callback
+{
+public:
+    int runKernel( InputArray _m1, InputArray _m2, OutputArray _model ) const
+    {
+        Mat m1 = _m1.getMat(), m2 = _m2.getMat();
+        const Point2f* from = m1.ptr<Point2f>();
+        const Point2f* to   = m2.ptr<Point2f>();
+
+        // we need 3 points to estimate affine transform
+        const int N = 6;
+        double buf[N*N + N + N];
+        Mat A(N, N, CV_64F, &buf[0]);
+        Mat B(N, 1, CV_64F, &buf[0] + N*N);
+        Mat X(N, 1, CV_64F, &buf[0] + N*N + N);
+        double* Adata = A.ptr<double>();
+        double* Bdata = B.ptr<double>();
+        A = Scalar::all(0);
+
+        for( int i = 0; i < (N/2); i++ )
+        {
+            Bdata[i*2] = to[i].x;
+            Bdata[i*2+1] = to[i].y;
+
+            double *aptr = Adata + i*2*N;
+            for(int k = 0; k < 2; ++k)
+            {
+                aptr[0] = from[i].x;
+                aptr[1] = from[i].y;
+                aptr[2] = 1.0;
+                aptr += N+3;
+            }
+        }
+
+        solve(A, B, X, DECOMP_SVD);
```

Comment:
For the full affine transform, the {a, b, c} model params only depend on the Xi outputs and the {d, e, f} params only depend on the Yi outputs.  So instead of solving one 6x6 problem, you can solve two 3x3 problems.  And the 3x3 problem can be solved analytically in the 2D case (since the last column is all ones).

```
        Mat M = _model.getMat();

        /*
         The estimate of {a, b, c} only depends on the Xi, and {d, e, f} only depends on the Yi
         
         solving AX = B
         | X1 |   | x1 y1 1 |   | a |
         | X2 | = | x2 y2 1 | * | b |
         | X3 |   | x3 y3 1 |   | c |

         | Y1 |   | x1 y1 1 |   | d |
         | Y2 | = | x2 y2 1 | * | e |
         | Y3 |   | x3 y3 1 |   | f |
         we solve each one analytically
         */
        double x1 = from[0].x;
        double y1 = from[0].y;
        double x2 = from[1].x;
        double y2 = from[1].y;
        double x3 = from[2].x;
        double y3 = from[2].y;
        
        double X1 = to[0].x;
        double Y1 = to[0].y;
        double X2 = to[1].x;
        double Y2 = to[1].y;
        double X3 = to[2].x;
        double Y3 = to[2].y;
        
        double d = 1. / ( x1*(y2-y3) + x2*(y3-y1) + x3*(y1-y2) );
        
        Mdata[0] = d * ( X1*(y2-y3) + X2*(y3-y1) + X3(y1-y2) );
        Mdata[1] = d * ( X1*(x3-x2) + X2*(x1-x3) + X3(x2-x1) );
        Mdata[2] = d * ( X1*(x2*y3 - x3*y2) + X2*(x3*y1 - x1*y3) + X3*(x1*y2 - x2*y1) );
        
        Mdata[3] = d * ( Y1*(y2-y3) + Y2*(y3-y1) + Y3(y1-y2) );
        Mdata[4] = d * ( Y1*(x3-x2) + Y2*(x1-x3) + Y3(x2-x1) );
        Mdata[5] = d * ( Y1*(x2*y3 - x3*y2) + Y2*(x3*y1 - x1*y3) + Y3*(x1*y2 - x2*y1) );
```

I haven't tested this, but the formulas came from Mathemetica.